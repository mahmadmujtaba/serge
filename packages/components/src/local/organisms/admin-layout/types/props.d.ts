import StatusBarTypes from '../../../molecules/statusbar/types/props'

export default interface PropTypes extends StatusBarTypes {
  /* half section title */
  type?: 'submit' | 'reset' | 'button'
  size?: 's' | 'm'
  onTabChange?: (tab: string, key: number, e: any) => void
  onPressBack?: () => void
  disabled?: boolean
  tabs?: Array<string>
  wargameChanged?: boolean
  activeTab?: string
}
