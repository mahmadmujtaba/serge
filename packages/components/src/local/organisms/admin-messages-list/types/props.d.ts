import { Message } from '@serge/custom-types'

export default interface PropTypes {
  /**
   * The messages array
   */
  messages: Array<Message>
  /*
   *  The function to perform to mark all messages as read
   */
  markAllAsRead: any
  /**
   * Current force that is seeing the messages
   */
  force: {
    name: string
    color: string
    icon?: string
  }
}
