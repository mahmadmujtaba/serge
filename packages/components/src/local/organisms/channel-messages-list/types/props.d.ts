import ChannelMessageProps from '../../../molecules/channel-message/types/props'
import ForcesInChannelProps from '../../../molecules/forces-in-channel/types/props'

interface MessagesTypes extends ChannelMessageProps {
  infoType?: boolean
  gameTurn?: number
}

export default interface PropTypes extends ForcesInChannelProps{
  /**
   * The list of channel messages properties required
   * for ChannelMessage components
   */
  messages: Array<MessagesTypes>
  /**
   * Callback on expanding message item
   */
  onRead?: { (detail: object): void }
}
