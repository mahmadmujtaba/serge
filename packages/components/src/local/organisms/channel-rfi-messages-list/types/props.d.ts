import ForcesInChannelProps from '../../../molecules/forces-in-channel/types/props'

import { MessageCustom, Role } from '@serge/custom-types'

export default interface PropTypes extends ForcesInChannelProps {
  isUmpire: boolean
  isRFIManager: boolean
  role: Role['name']
  messages: MessageCustom[]
  onChange: (nextMessage: MessageCustom) => void
  onRead: (nextMessage: MessageCustom) => void
}
