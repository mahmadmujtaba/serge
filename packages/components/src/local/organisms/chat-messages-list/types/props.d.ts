import ChatMessageProps from '../../../molecules/chat-message/types/props'
import ForcesInChannelProps from '../../../molecules/forces-in-channel/types/props'

export default interface PropTypes extends ForcesInChannelProps {
  /**
   * The list of channel messages properties required
   * for ChatMessage components
   */
  messages: Array<ChatMessageProps>
}
