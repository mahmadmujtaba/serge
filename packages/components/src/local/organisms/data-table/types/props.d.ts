import React from 'react'

export type RowDataType = string | { component: React.ReactElement | null, label: string }
export type RowWithCollapsibleType = {
  // collapsible is a function witch one will return a ReactElement and will be rendered only on call
  // cb: is a callback function with one should fire some logick on collapsible content change, ex close collapsible component
  collapsible: (cb?: () => void) => React.ReactElement
  cells: Array<RowDataType>
}
export type RowType = Array<RowDataType | RowWithCollapsibleType>
export default interface Props {
  /**
   * List of columns
   */
  columns: Array<string | { filters: Array<string>, label: string }>
  /**
   * List of data
   */
  data: Array<RowType>
}
