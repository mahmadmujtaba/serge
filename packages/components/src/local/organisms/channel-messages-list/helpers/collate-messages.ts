import { MessagesTypes } from '../types/props'
import { INFO_MESSAGE_CLIPPED, UMPIRE_FORCE } from '@serge/config'
import { MessageChannel, MessageCustom } from '@serge/custom-types'
import moment from 'moment'


/** Generate the Serge name for the supplied cell (1,3) == "B03" */
const collateMessages = (messages: MessageChannel[], 
     selectedForceId: string, 
     onRead?: { (detail: object): void }): MessagesTypes[]  => {

return messages.map((item: MessageChannel | any) => {
      if(item.messageType === INFO_MESSAGE_CLIPPED) {
        // this is a turn marker, we return it unchanged
        // TODO: I think this would return wargame documents too. 
        // we probably need to reduce it to just the game turn. Then
        // we reduce the list of messages to remove duplicate game turns
        return item
      } else {
        const messageItem: MessageCustom = item
        const {
          details,
          message,
          isOpen,
          hasBeenRead,
          infoType,
          gameTurn
        } = messageItem
        const { role, forceColor } = details.from || {}
        const { messageType, privateMessage } = details
        const dynamicBorderColor = `${forceColor}${hasBeenRead ? 'B3':''}`
        const timestamp = moment(details.timestamp)
        const isUmpire = selectedForceId === UMPIRE_FORCE
        const detail = message || {}
        let title
        if (detail.title) {
          title = detail.title
        } else if (detail.Title) { 
          title = detail.Title // have Title field in RFI format
        } else if(detail.content) {
          // yes, we have content (probably chat) use it
          title = detail.content
        } else {
          // no content, just use message-type
          title = details.messageType
        }
        return {
          ...item,
          borderColor: dynamicBorderColor,
          infoType,
          gameTurn,
          isOpen,
          title,
          timestamp,
          role,
          forceColor,
          messageType,
          hasBeenRead,
          privateMessage,
          isUmpire,
          detail,
          onRead
        }
      }
    })
}

export default collateMessages
