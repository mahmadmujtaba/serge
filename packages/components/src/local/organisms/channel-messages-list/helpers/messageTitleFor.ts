import { MessageCustom } from '@serge/custom-types'
export default function messageTitleFor(message: MessageCustom): string {
    if (message.message.title) {
        return message.message.title
    } else if (message.message.Title) {
        return  message.message.Title // have Title field in RFI format
    } else if (message.message.content) {
        // yes, we have content (probably chat) use it
        return  message.message.content
    } else {
        // no content, just use message-type
        return  message.details.messageType
    }
}

