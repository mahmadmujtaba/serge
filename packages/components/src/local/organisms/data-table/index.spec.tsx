/* global it expect */
import React from 'react'
import renderer from 'react-test-renderer'
import Badge from '../../atoms/badge'
import RfiForm from '../../molecules/rfi-form'
import { MessageCustom } from '@serge/custom-types/message'
import { GameMessagesMockRFI } from '@serge/mocks'
import { mostRecentOnly } from '@serge/helpers'

import DataTable, { ROW_DATA_TYPE, ROW_WITH_COLLAPSIBLE_TYPE } from './index'
import { RowWithCollapsibleType } from './types/props'

const rfiMessages = (GameMessagesMockRFI as MessageCustom[]).filter((message: MessageCustom) => message.details.messageType === 'RFI')
const newest = mostRecentOnly(rfiMessages) as MessageCustom[]

const rfiData: any[][] = []

const data: RowWithCollapsibleType[] = newest.map((message, rowIndex) => {
  const messageItem = (message as MessageCustom)
  const row = [
    messageItem._id,
    messageItem.details.channel,
    messageItem.details.from.roleName,
    messageItem.details.from.forceColor,
    messageItem.message.Title,
      messageItem.details.collaboration?.status,
      messageItem.details.collaboration?.owner
  ]

  rfiData.push(row)

  const statusColors = {
    Unallocated: '#B10303',
    'In progress': '#E7740A',
    'Pending review': '#434343',
    Released: '#007219'
  }
  return {
    type: ROW_WITH_COLLAPSIBLE_TYPE,
    rowKey: 'rowKey' + rowIndex,
    collapsible: (): React.ReactElement => (
      <RfiForm onSubmit={console.log} onReject={console.log} message={(newest[rowIndex] as MessageCustom)} />
    ),
    cells: [
      {
        rowkey: 'id',
        type: ROW_DATA_TYPE,
        component: null,
        label: messageItem._id
      },
      {
        rowkey: 'channel',
        type: ROW_DATA_TYPE,
        component: null,
        label: messageItem.details.channel
      },
      {
        rowkey: 'roleName',
        type: ROW_DATA_TYPE,
        component: <Badge customBackgroundColor={messageItem.details.from.forceColor} label={messageItem.details.from.roleName}/>,
        label: messageItem.details.from.roleName
      },
      {
        rowkey: 'Title',
        type: ROW_DATA_TYPE,
        component: null,
        label: messageItem.message.Title
      },
      {
        rowkey: 'Status',
        type: ROW_DATA_TYPE,
        component: <Badge customBackgroundColor={messageItem.details.collaboration?.status ? statusColors[messageItem.details.collaboration?.status] : '#434343'} customSize="large" label={messageItem.details.collaboration?.status}/>,
        label: `${messageItem.details.collaboration?.status}`
      },
      {
        rowkey: 'Owner',
        type: ROW_DATA_TYPE,
        component: messageItem.details.collaboration?.owner ? <Badge customBackgroundColor="#434343" label={messageItem.details.collaboration?.owner.roleName} /> : null,
        label: ''
      }
    ]
  }
})

const uniqueFieldValues = (messages: any[], col: number): any => {
  // find items with unique items in set column
  const uniqueValues = messages.filter((elem, index) => rfiData.findIndex(obj => obj[col] === elem[col]) === index)
  // produce array with just field of interest
  const values = uniqueValues.map((item: any) => item && item[col])
  // swap undefined for string
  return values.map((item: any) => item === undefined ? '[Undefined]' : item)
}

const columns = [
  'ID',
  {
    filters: uniqueFieldValues(rfiData, 1),
    label: 'Channel'
  },
  {
    filters: uniqueFieldValues(rfiData, 2),
    label: 'From'
  },
  'Title',
  {
    filters: [
      'Unallocated',
      'In Progress',
      'Pending Review',
      'Released'
    ],
    label: 'Status'
  },
  {
    filters: uniqueFieldValues(rfiData, 6),
    label: 'Owner'
  }
]

describe('DataTable component:', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <DataTable columns={columns} data={data} />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
