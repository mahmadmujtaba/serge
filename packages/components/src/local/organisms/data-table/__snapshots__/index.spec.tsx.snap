// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable component: renders correctly 1`] = `
<div
  className="MuiTableContainer-root"
>
  <table
    className="MuiTable-root makeStyles-table-1"
    role={null}
  >
    <thead
      className="MuiTableHead-root makeStyles-tableHead-2"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span>
            First column
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span>
            Second column
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span>
            Third column
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root makeStyles-tableBody-3"
      role={null}
    >
      <tr
        className="MuiTableRow-root makeStyles-tableRow-4 "
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <svg
            className="svg-inline--fa fa-plus"
          />
          Â 
          Row 1 Cell 1
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          Row 1 Cell 2
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          Row 1 Cell 3
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
